bpf/lib/nodeport.h

```diff
@@ -449,7 +449,7 @@ int tail_nodeport_ipv6_dsr(struct __ctx_buff *ctx)
 		}
 	}
 
-	return ctx_redirect(ctx, fib_params.l.ifindex, 0);
+	return ctx_redirect(ctx, DIRECT_ROUTING_DEV_IFINDEX, 0);
```

Comment:
Why is this change required?

Reply:
@brb For example, when there is a bridge br0 with eth0 attached, IP is set on br0 (for fib_lookup), and eth0 is a L2 device without IP, we should use br0's ifindex for fib_lookup, but still forward packets use eth0. In this case DIRECT_ROUTING_DEV_IFINDEX = eth0, FIB_LOOKUP_DEV_IFINDEX = br0.

Reply:
Sorry for taking long on this, and thanks for the explanation.

Anyway, I'm thinking maybe for the sake of the datapath simplicity, we could omit the `BPF_FIB_LOOKUP_DIRECT` flag in the `fib_lookup()`, and set ifindex=0. The downside of this approach is that might introduce a performance penalty. Would you mind measuring it?

If the suggested approach is OK, another thing is that for the DSR mode we don't actually need the LB IP addr in the datapath. So, maybe we could even get rid of the `--bpf-lb-dev-ip-addr-inherit` flag which would simplify the code by a bit.

Reply:
@brb in fact I tried omitting the BPF_FIB_LOOKUP_DIRECT flag and setting ifindex = 0. However, it didn't work because the kernel helper function fib_lookup() doesn't support ifindex = 0. It validates that ifindex if valid and can get dev from it. I think the kernel helper should be fixed to avoid this validation when BPF_FIB_LOOKUP_DIRECT is not used. Until the kernel helper is fixed, this patch seems to be the only way to make it work.

For --bpf-lb-dev-ip-addr-inherit, I think the LB IP is still needed for DSR mode probably for environments where source IP range is checked in routers with RPF settings.